synapse/lib/crypto/vault.py

```diff
@@ -0,0 +1,516 @@
+        #return (iden, data, rkey.sign(data + cert[0]))
+
+    def verify(self, byts, sign):
+        '''
+        Verify that the the Cert signed a set of bytes.
+
+        Args:
+            byts (bytes): Data to check.
+            sign (bytes): Signature to verify.
+
+        Returns:
+            bool: True if the Cert signed the byts, False otherwise.
+        '''
+        return self.rpub.verify(byts, sign)
+
+    def signed(self, cert):
+        '''
+        Check if this cert signed the given Cert and return the info.
+
+        Args:
+            cert (Cert): A Cert to confirm that we signed.
+
+        Returns:
+            dict: The signer info dict ( or None if not signed ).
+        '''
+        byts = cert.toknbytes()
+
+        for iden, data, sign in cert.signers():
+
+            if iden != self.iden():
+                continue
+
+            if self.verify(data + byts, sign):
+                return s_msgpack.un(data)
+
+    def save(self):
+        '''
+        Serialize the certificate to bytes for storage.
+
+        Returns:
+            bytes: A msgpack encoded form of the Cert.
+        '''
+        return s_msgpack.en(self.cert)
+
+    @staticmethod
+    def load(byts, rkey=None):
+        '''
+        Create a Cert object from the bytes.
+
+        Args:
+            byts (bytes): Bytes from a previously saved Cert
+            rkey (s_rsa.PriKey): The RSA Private Key assocaited with the Cert.
+
+        Returns:
+            Cert: A Cert object for the bytes and RSA private key.
+        '''
+        return Cert(s_msgpack.un(byts), rkey=rkey)
+
+class Vault(s_eventbus.EventBus):
+
+    '''
+    tokn:
+        {
+            'user': <str>,
+            'rsa:pub': <bytes>,
+        }
+
+    cert:
+        ( <toknbyts>, {
+            "signers": (
+                <sig>,
+            ),
+        })
+
+    sig:
+        # NOTE: <iden> must *only* be used for pub key lookup
+        (<iden>, <bytes(signdata)>, <signbytes>),
+
+    '''
+    def __init__(self, path):
+
+        s_eventbus.EventBus.__init__(self)
+
+        self.kvstor = s_kv.KvStor(path)
+        self.onfini(self.kvstor.fini)
+
+        self.info = self.kvstor.getKvLook('info')  # Internal Housekeeping
+        self.keys = self.kvstor.getKvLook('keys')  # iden -> private RSA keys
+        self.certs = self.kvstor.getKvLook('certs')  # iden -> signed Cert
+        self.roots = self.kvstor.getKvLook('roots')  # iden -> CAs capable of signing
+        self.certkeys = self.kvstor.getKvLook('keys:bycert')  # cert -> private RSA keys
+        self.usercerts = self.kvstor.getKvLook('certs:byuser')  # user -> certs
+
+    def genRsaKey(self):
+        '''
+        Generate a new RSA key and store it in the vault.
+
+        Returns:
+            s_rsa.PriKey: The new RSA key.
+        '''
+        rkey = s_rsa.PriKey.generate()
+
+        iden = rkey.iden()
+        self.keys.set(iden, rkey.save())
+
+        return rkey
+
+    def setRsaKey(self, name, byts):
```

Comment:
This API is busted - is it intended that we want to replace arbitrary user certs?

Reply:
I think we should remove this method